### ISSUE TITLE
Content offset a couple lines

### ISSUE THREAD

**Issue by ianpan870102** [akermu/emacs-libvterm#229]
After the commit titled "Fix #226 #198 #227", the cursor position seems to be correct, but there are always  **a few lines of offset** (as if someone scrolled down explicitly a couple of lines) that messes up the content. 
_Please let me know if this description is unclear, I can provide some screenshots._

**Comment by jixiuf** [comment]
screenshots please .

**Comment by ianpan870102** [comment]
Sure! Screenshots with explanation:

Here's a brand new vterm buffer with the command "ls -la", this is how it was supposed to look like:
<img width="652" alt="Screenshot 2020-02-04 at 17 19 42" src="https://user-images.githubusercontent.com/36686900/73730899-8d425580-4772-11ea-834e-34a0690f8741.png">

However, vterm seems to automatically scroll down a few lines, to become so:
<img width="653" alt="Screenshot 2020-02-04 at 17 18 51" src="https://user-images.githubusercontent.com/36686900/73730955-9fbc8f00-4772-11ea-9f0c-cf2b57edb801.png">

Another example would be "top -u". Here's how it should look like:
<img width="1440" alt="Screenshot 2020-02-04 at 17 21 06" src="https://user-images.githubusercontent.com/36686900/73731079-ced30080-4772-11ea-94f8-a36b1c12647e.png">

But this is how it's presented in emacs vterm, again with the top couple of lines missing (the sumary title, PID, %CPU etc.), as if someone scrolled down by hand.
<img width="1440" alt="Screenshot 2020-02-04 at 17 20 51" src="https://user-images.githubusercontent.com/36686900/73731140-ec07cf00-4772-11ea-8f9b-cee06932b6ab.png">

**Comment by jixiuf** [comment]
can you reproduced this with `make run` in  emacs-libvterm/build/

**Comment by jixiuf** [comment]
65b76502d6503e197662f506d6d4212fb333b972 should fix this problem 
you set 
```
  (setq-default line-spacing 3)
```
https://github.com/ianpan870102/.personal-emacs.d/blob/master/init.el#L60

**Comment by ianpan870102** [comment]
I really appreciate your time, even going into my personal config to help me find the problem! After your commit with `(setq-local line-spacing nil)`, the content scroll-off problem disappears and everything's smooth. I'll let you know if similar problems resurface, but I'll close this issue for now. Again, thanks for making this package happen, it's a wonderful terminal emulation.

**Comment by ianpan870102** [comment]
Hi @jixiuf sorry to break this to you, but the problem with "a few lines of offset" seems to have reappeared.

## Here are a couple of screenshots to illustrate the issue:

### How my screen looks after executing the "clear" command.
![image](https://user-images.githubusercontent.com/36686900/86216633-ecd14f80-bbb0-11ea-8f1c-208869941dd2.png)

### How my buffer looks when running "top". (Notice the first 4 lines of content is wrong)

![image](https://user-images.githubusercontent.com/36686900/86216693-083c5a80-bbb1-11ea-8e8c-e1b47d775f71.png)

# My investigation: found a pattern

I tried to find a pattern and here's what I got: the vterm buffer will behave normally (i.e. clear the screen cleanly) **if the bottom of terminal buffer is never physically reached**. That is, if we always remember to **clear** screen before the terminal buffer content touches the bottom edge, we're good. But as soon as we have one or more command(s) that prints content for more than 1 buffer height (page), then **clear** screen no longer works as expected, leaving the uncleared couple of lines on top.

# Other notes

I remembered to set line-spacing to nil, and I'm also using a monospace font (Consolas to be precise).

**Comment by ianpan870102** [comment]
@Sbozzolo @jixiuf Friendly ping.

**Comment by Sbozzolo** [comment]
Thanks, I can reproduce with the following steps:
1. Get a vterm buffer with full scrollback (you can set `vterm-max-scrollback` a small number)
2. Press `RET` until the prompt reaches the end of the buffer
3. Fire up `top`.

In my tests, the problem is not there if the scrollback is full. Can you please verify this?

**Comment by ianpan870102** [comment]
Hi @Sbozzolo thanks for reaching back. My vterm-max-scrollback value is the default 1000 (checked with C-h v). 

>In my tests, the problem is not there if the scrollback is full. Can you please verify this?

Could you please explain how I can "verify the issue when the scrollback is full"? I don't quite understand what you meant. Cheers!

**Comment by Sbozzolo** [comment]
> Could you please explain how I can "verify the issue when the scrollback is full"? I don't quite understand what you meant. Cheers!

Yes, if `vterm-max-scrollback` is set to 1000, it means that 1000 lines can be printed on screen before `vterm` starts removing the oldest ones to make room for new lines. I found that the bug you reported is not triggered with a fresh `vterm` that has seen fewer than 1000 lines.

To test this, just open a new vterm, ls a couple of times, and try to reproduce the bug.

**Comment by ianpan870102** [comment]
I see what you mean now, thanks!
Actually, the bug I reported occurs once "the bottom of the buffer is reached"; it doesn't have to be > 1000 lines. That is, I can reproduce it just be opening a new vterm buffer, `ls` a couple of times until it reaches the end of buffer page (which is around 40 lines?), then `clear` or `top` to trigger the offset bug.

**Comment by ianpan870102** [comment]
### Upon further investigation

The problem seems to stem from my `line-spacing` definition in my `init.el`. I prefer to `(setq-default line-spacing 3)`, and apparently setting vterm's local `line-spacing` value to `nil` isn't enough -- the global line-spacing value must be `nil` too. This is the only way that the bug ceased to show up on my end.

**Comment by pjones** [comment]
I'm seeing something similar.  When I `clear` the screen I still have a small number of lines at the top of the terminal from previous commands.  Calling `vterm-clear-scrollback` followed by `clear` fixes the problem.

I have `vterm-max-scrollback` set to 10,000 and `line-spacing` is the default (`nil`).

Otherwise I absolutely love vterm.  Thank you for working on this project @Sbozzolo!

**Comment by ianpan870102** [comment]
> Calling vterm-clear-scrollback followed by clear fixes the problem.

Thanks for the tip! In which case, perhaps we can bind vterm-clear to simly vterm-clear + vterm-clear-scrollback? Btw the manual keybinding for that would be `C-l` followed by `C-c C-l`. (Or even better: hold down `Ctrl` and type `l c l`)

**Comment by ianpan870102** [comment]
To use `C-l` to clear screen 'and' scroll back, here's the easy solution I came up with:

```emacs-lisp
(define-key vterm-mode-map (kbd "C-l") #'(lambda ()
                                           (interactive)
                                           (vterm-clear)
                                           (vterm-clear-scrollback))))
```

**Comment by pjones** [comment]
@ianpan870102 the only problem with that solution is that I don't want to delete the scrollback buffer as I use it quite often.  I'd rather have a more permanent solution that prevents this from happening in the first place.

**Comment by ianpan870102** [comment]
@pjones Fair enough, let's leave this issue open and hopefully we can see a fix in the future :)

**Comment by Just-Insane** [comment]
I think I'm running into the same issue. I'm currently only noticing it in tmux, but when I first attach to a tmux session or `clear` the terminal, everything looks fine. However, every time I press enter or issue a command from that state, the tmux status bar at the top goes off the screen.

`top`/`htop` appear to display fine however.

I've tried the workaround of setting `(setq-default line-spacing nil)` and `(setq-local line-spacing nil)` in `init.el` and `.spacemacs` respectively, and the issue persists.
